scripts/utils/Translator/Translator.ts

```diff
@@ -33,6 +34,58 @@ abstract class Translator {
     private trimForLogs(text: string) {
         return `${text.slice(0, 80)}${text.length > 80 ? '...' : ''}`;
     }
+
+    /**
+     * Validate that placeholders are all present and unchanged before and after translation.
+     */
+    public validateTemplatePlaceholders(original: string, translated: string): boolean {
+        const extractPlaceholders = (s: string) =>
+            Array.from(s.matchAll(/\$\{[^}]*}/g))
+                .map((m) => m[0])
+                .sort();
+        const originalSpans = extractPlaceholders(original);
+        const translatedSpans = extractPlaceholders(translated);
+        return JSON.stringify(originalSpans) === JSON.stringify(translatedSpans);
+    }
+
+    /**
+     * Validate that the HTML structure is the same before and after translation.
+     */
+    public validateTemplateHTML(original: string, translated: string): boolean {
+        // Attributes that are allowed to be translated
+        const TRANSLATABLE_ATTRIBUTES = ['alt', 'title', 'placeholder', 'aria-label', 'aria-describedby', 'aria-labelledby', 'value'];
+
+        const parseHTMLStructure = (s: string) => {
+            const doc = parseDocument(s);
+            const elements = DomUtils.getElementsByTagName(() => true, doc, true);
+
+            return elements.map((element) => {
+                const tagName = element.name.toLowerCase();
+
+                // Extract attributes, excluding translatable ones
+                const attributes: string[] = [];
+                if (element.attribs) {
+                    Object.entries(element.attribs).forEach(([attrName, attrValue]) => {
```

Comment:
NAB because we haven't standardized, but I've been trying to push us to start using plain loops rather than `forEach`, `map`, `reduce`, _particularly in scripts_, because it's more performant, flexible, and imo more readable:

```ts
const attributes: string[] = [];
for (const [attrName, attrValue] of Object.entries(element.attribs ?? {})) {
    const normalizedAttrName = attrName.toLowerCase();
    if (!TRANSLATABLE_ATTRIBUTES.includes(normalizedAttrName)) {
        attributes.push(`${normalizedAttrName}="${attrValue ?? ''}"`);
    }
}

return {
    tagName,
    attributes: attributes.sort(),
};
```